### ISSUE TITLE
Port System.Net.TransportContext missing members

### ISSUE THREAD

**Issue by ericeil** [dotnet/corefx#12136]
```csharp
    public abstract class TransportContext {
        public virtual IEnumerable<TokenBinding> GetTlsTokenBindings();
    }
```

**Comment by geoffkizer** [comment]
This would require adding the TokenBinding class.  TokenBinding has an internal constructor, and is only used in HttpListener (see HttpListenerRequestContext).  I'm not sure what the implications of making the constructor public are.

@davidsh, do you know what this does, exactly?  Do we actually need it for netstandard 2.0?

**Comment by davidsh** [comment]
cc: @CIPop

**Comment by CIPop** [comment]
Token binding is a client and server new TLS feature: https://tools.ietf.org/html/draft-popov-token-binding-00
/cc @ericeil In .Net Desktop support for TokenBinding is available for HttpListener since .Net 4.6.1. From the client-side, at least IE/Edge and Chrome have support for this security feature.

**Comment by geoffkizer** [comment]
@weshaggard, @stephentoub 

The problem here is adding the TokenBinding class.  It has only an internal constructor today.  To make this usable (e.g. by HttpListener) we would need to add a public constructor to it.  Is that acceptable to do?

**Comment by stephentoub** [comment]
How about if we add it to System.Net.HttpListener (or whatever assembly @ericeil is putting HttpListener in), use the internal ctor, and then if we find a need in the future we can type forward it down, add appropriate additional API surface area, etc.  Reasonable?

**Comment by davidsh** [comment]
>The problem here is adding the TokenBinding class. It has only an internal constructor today. 

The reason for that is that devs are not to create this on their own.  This object is returned by other code.

**Comment by geoffkizer** [comment]
@stephentoub I don't think that works.

TransportContext already exists, and lives in System.Net.Primitives.  It's used by System.Net.Security and System.Net.Http.

TransportContext is missing the GetTlsTokenBindings method.  This method returns an IEnumerable<TokenBinding>.

HttpListener is the only code that actually implements/uses GetTlsTokenBindings.  But other code uses TransportContext, just not the GetTlsTokenBindings method.

so, TokenBinding <= TransportContext (in Primitives today) <= System.Net.Security,  System.Net.Http, System.Net.HttpListener.

**Comment by stephentoub** [comment]
Ah, I see, I was missing the necessary context (no pun intended).

**Comment by geoffkizer** [comment]
@davidsh How would we implement HttpListenerRequestContext -- which derives from TransportContext and overrides GetTlsTokenBindings?

**Comment by CIPop** [comment]
Another option: Make the relevant APIs virtual and return a derived class that inherits from `System.Net.Primitives`/`TokenBinding` and is implemented in the internal `System.Net.HttpListener`/`System.Net.Internals.TokenBinding`.
The virtual modifier will be required for `GetRawTokenBindingId` and `BindingType`.

**Comment by CIPop** [comment]
@geoffkizer No... you're right: it doesn't work without the protected ctor. 
Looking closer at the API surface, the fact that the ctor is not protected is most likely a bug: `TokenBinding` has been designed after `ChannelBinding` which has one.

We should add the protected ctor although I'm not sure how that fits in the netstandard20 architecture. 
@weshaggard or @ericstj might know if this is possible or not.

**Comment by geoffkizer** [comment]
If we're adding a constructor anyway, we may as well make it public and not have to make the other methods virtual.

**Comment by davidsh** [comment]
The constructor for TransportContext should be protected and not public.  It is an abstract class.

**Comment by geoffkizer** [comment]
@davidsh It's not TransportContext that needs a constructor.  It's TokenBinding.

**Comment by weshaggard** [comment]
Based on what I'm seeing there are only a few options:

1. We remove these APIs from the standard
2. We combine more of these types by moving them down into System.Net.Primitives
3. We add the new APIs as netcoreapp specific so we can use them in our implementation but won't be part of netstandard (at least not until some future version if we want to add them to the standard).

I don't have strong opinion about removing option (1), the usage data doesn't point at much usage so that could be an option but I'd look to you guys to say whether or not that would block any interesting scenarios. Beyond that option (3) seems like it would be reasonable, but would require some work to convert the libraries in question to be netcoreapp specific.

**Comment by CIPop** [comment]
> @davidsh It's not TransportContext that needs a constructor. It's TokenBinding.

Adding a public ctor isn't how we've designed the object model with the API review board the partner team that implemented the feature. The key point was that we wanted offer a similar developer experience to the existing `ChannelBinding` object model.

> We add the new APIs as netcoreapp specific so we can use them in our implementation but won't be part of netstandard (at least not until some future version if we want to add them to the standard).

I think this is a good approach. (Which I believe will include 1 as well.)

**Comment by geoffkizer** [comment]
> Adding a public ctor isn't how we've designed the object model with the API review board the partner team that implemented the feature. The key point was that we wanted offer a similar developer experience to the existing ChannelBinding object model.

I don't understand this.  HttpListener calls `new TokenBinding(...)`, see here: https://referencesource.microsoft.com/#System/net/System/Net/HttpListenerRequest.cs,1111

Since these now live in separate assemblies, it seems like the most straightforward thing to do is to just add a public ctor.

**Comment by geoffkizer** [comment]
> I think this is a good approach. (Which I believe will include 1 as well.)

What does this mean in practice?

**Comment by stephentoub** [comment]
> I don't understand this. HttpListener calls new TokenBinding(...), see here: https://referencesource.microsoft.com/#System/net/System/Net/HttpListenerRequest.cs,1111

Presumably HttpListener would just provide its own derived type.  I don't have a strong opinion here, but if a TokenBinding is not meant to be created by arbitrary code, and it seems like that's the case and why it has an internal rather than public ctor today, then I can understand the push to have a protected instead of public ctor.  Making it protected allows for code to really create one if it needs to, but it avoids accidentally misusing it by creating one when doing so wasn't a good idea... in other words, it doesn't prevent the case, but it makes it more difficult, and thus helps to thwart misuse.  That said, I have no context on the type or its intended usage.

**Comment by geoffkizer** [comment]
Given that this is proving to be tricky to actually implement, I think we should seriously consider just cutting this for now.  We can always revisit in a future release.

**Comment by CIPop** [comment]
> What does this mean in practice?

I'm adding some documentation here: #13026 on how to add netcoreapp1.1. I also plan to submit a PR soon with the code changes.

**Comment by geoffkizer** [comment]
Thanks @CIPop, that documentation is helpful.

**Comment by geoffkizer** [comment]
Per discussion, we will cut this.

I'll file a separate issue against netstandard2.0.
